pygen/pygen_src/riscv_asm_program_gen.py

```diff
@@ -577,14 +586,59 @@ def gen_signature_handshake(self, instr, signature_type, core_status = "INITIALI
         pass
 
     def add_directed_instr_stream(self, name, ratio):
-        pass
+        self.directed_instr_stream_ratio[name] = ratio
+        logging.info("Adding directed instruction stream:%0s ratio:%0d/1000", name, ratio)
 
     def get_directed_instr_stream(self):
-        pass
-
-    def generate_directed_instr_stream(self, hart = 0, label = "", original_instr_cnt = None,
+        opts = []
+        for i in range(cfg.max_directed_instr_stream_seq):
+            arg = "directed_instr_{}".format(i)
+            stream_name_opts = "stream_name_{}".format(i)
+            stream_freq_opts = "stream_freq_{}".format(i)
+            if(arg in args):
+                val = args_dict[arg]
+                opts = val.split(",")
+                if(len(opts) != 2):
+                    logging.critical(
+                        "Incorrect directed instruction format : %0s, expect: name,ratio", val)
+                else:
+                    self.add_directed_instr_stream(opts[0], int(opts[1]))
+            elif(stream_name_opts in args and stream_freq_opts in args):
+                stream_name = args_dict[stream_name_opts]
+                stream_freq = args_dict[stream_freq_opts]
+                self.add_directed_instr_stream(stream_name, stream_freq)
+
+    def generate_directed_instr_stream(self, hart = 0, label = "", original_instr_cnt = 0,
                                        min_insert_cnt = 0, kernel_mode = 0, instr_stream = []):
-        pass
+        new_instr_stream = riscv_rand_instr_stream()
+        instr_insert_cnt = 0
+        idx = 0
+        if(cfg.no_directed_instr):
+            return
+        for instr_stream_name in self.directed_instr_stream_ratio:
+            instr_insert_cnt = int(original_instr_cnt *
+                                   self.directed_instr_stream_ratio[instr_stream_name] // 1000)
+            if(instr_insert_cnt <= min_insert_cnt):
+                instr_insert_cnt = min_insert_cnt
+            logging.info("Insert directed instr stream %0s %0d/%0d times",
+                         instr_stream_name, instr_insert_cnt, original_instr_cnt)
+            for i in range(instr_insert_cnt):
+                name = "{}_{}".format(instr_stream_name, i)
+                object_h = factory(instr_stream_name)
+                object_h.name = name
+                if(object_h is None):
+                    logging.critical("Cannot create instr stream %0s", name)
+                new_instr_stream = copy.copy(object_h)
+                if(new_instr_stream is not None):
```

Comment:
Could you please change this to `if object_h`? `is not None` seems redundant.

Reply:
fixed